submission/races.go

```diff
@@ -0,0 +1,299 @@
+	state *safeSubmissionState, submitter Submitter) groupState {
+	groupURLs := make([]string, 0, len(group.LogURLs))
+	for logURL := range group.LogURLs {
+		groupURLs = append(groupURLs, logURL)
+	}
+
+	type count struct{}
+	counter := make(chan count, len(groupURLs))
+
+	// Randomize the order in which we send requests to the logs in a group
+	// so we maximize the distribution of logs we get SCTs from.
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		subCtx, cancel := context.WithCancel(ctx)
+		go func(i int, logURL string) {
+			timeoutchan := time.After(submitInterval(i, parallelStart, SubmitBatchInterval))
+			select {
+			case <-subCtx.Done():
+				counter <- count{}
+				return
+			case <-timeoutchan:
+			}
+			if state.groupComplete(group.Name) {
+				cancel()
+				counter <- count{}
+				return
+			}
+			if firstRequested := state.request(logURL, cancel); !firstRequested {
+				counter <- count{}
+				return
+			}
+			sct, err := submitter.SubmitToLog(subCtx, logURL, chain)
+			// TODO(Mercurrent): verify SCT
+			state.handleResult(logURL, sct, err)
+			counter <- count{}
+		}(i, groupURLs[urlNum])
+	}
+	// Wait until either all logs within groups are processed or context is cancelled.
+	for i := 0; i < len(groupURLs); i++ {
+		select {
+		case <-ctx.Done():
+			return groupState{Name: group.Name, Success: state.groupComplete(group.Name)}
+		case <-counter:
+			if state.groupComplete(group.Name) {
+				return groupState{Name: group.Name, Success: true}
+			}
+		}
+	}
+	return groupState{Name: group.Name, Success: state.groupComplete(group.Name)}
+}
+
+func parallelCalls(groups ctpolicy.LogPolicyData) map[string]int {
+	nums := make(map[string]int)
+	var subsetSum int
+	for _, g := range groups {
+		incl := g.MinInclusions
+		if incl < 0 {
+			incl = 0
+		}
+		nums[g.Name] = incl
+		if !g.IsBase {
+			subsetSum += incl
+		}
+	}
+	if _, hasBase := nums[ctpolicy.BaseName]; hasBase {
+		if nums[ctpolicy.BaseName] >= subsetSum {
+			nums[ctpolicy.BaseName] -= subsetSum
+		} else {
+			nums[ctpolicy.BaseName] = 0
+		}
+	}
+	return nums
+}
+
+// AssignedSCT represents SCT with logURL of log-producer.
+type AssignedSCT struct {
+	LogURL string
+	SCT    *ct.SignedCertificateTimestamp
+}
+
+func completenessError(groupComplete map[string]bool) error {
+	failedGroups := []string{}
+	for name, success := range groupComplete {
+		if !success {
+			failedGroups = append(failedGroups, name)
+		}
+	}
+	if len(failedGroups) > 0 {
+		return fmt.Errorf("log-group(s) %s didn't receive enough SCTs", strings.Join(failedGroups, ", "))
+	}
+	return nil
+}
+
+// GetSCTs picks required number of Logs according to policy-group logic and
+// collects SCTs from them.
+// Emits all collected SCTs even when unable-to-complete-error produced.
+func GetSCTs(ctx context.Context, submitter Submitter, chain []ct.ASN1Cert, groups ctpolicy.LogPolicyData) ([]*AssignedSCT, error) {
```

Comment:
This function is the essence of this package, I'd expect it to be on the top.

Reply:
I don't mind moving funcs around, but could you provide any Go-style recom on that? I see some standard libs contradicting this policy. Probably those are old or not style-concerned though.

Reply:
That's my personal taste and feeling, in particular when reading this PR, so if you prefer to not change it I'm not insisting. My taste was partially inspired by Cloud API [guidelines](https://cloud.google.com/apis/design/file_structure) where they recommend putting service definition (i.e. what matters most) first, and the details below.